securedrop_client/logic.py

```diff
@@ -129,40 +130,66 @@ def call_api(self, function, callback, timeout, *args, **kwargs):
         timeout signal. Any further arguments are passed to the function to be
         called.
         """
+
         if not self.api_thread:
+            self.timer = QTimer()
+            self.timer.timeout.connect(lambda: self.timeout_api_call.emit())
+            self.timer.setSingleShot(True)
+            self.timer.start(20000)
+
             self.api_thread = QThread(self.gui)
             self.api_runner = APICallRunner(function, *args, **kwargs)
             self.api_runner.moveToThread(self.api_thread)
-            self.api_runner.call_finished.connect(callback)
-            self.api_runner.timeout.connect(timeout)
-            self.finish_api_call.connect(self.api_runner.on_cancel_timeout)
+
+            # handle successful call: copy response data, reset the
+            # client, give the user-provided callback the response
+            # data
+            self.api_runner.call_finished.connect(lambda r: self.successful_api_call(r, callback))
+
+            # we've started a timer. when that hits zero, call our
+            # timeout function
+            self.timeout_api_call.connect(lambda: self.timeout_cleanup(timeout))
+
+            # when the thread starts, we want to run `call_api` on `api_runner`
             self.api_thread.started.connect(self.api_runner.call_api)
-            self.api_thread.finished.connect(self.call_reset)
+
             self.api_thread.start()
 
+        else:
+            logger.info("There's already an API request running, so I'm not going to ignore this one (XXX this may not be the coolest thing to do...)")
+
     def call_reset(self):
         """
         Clean up this object's state after an API call.
         """
         if self.api_thread:
-            self.finish_api_call.emit()
+            self.timeout_api_call.disconnect()
             self.api_runner = None
             self.api_thread = None
+            self.timer = None
```

Comment:
`call_reset` is called by our callback wrappers. We take care to disconnect the timeout handler callbacks: if we don't, then we'll just add more and more callbacks over time, and when one call finally times out, *all* collected timeout handlers will get triggered (even for requests that have long-since completed).